### ISSUE TITLE
vite with SSR: require is not defined

### ISSUE THREAD

**Issue by omridevk** [module-federation/vite#156]
when using SSR, I am getting the following error:
```
11:02:24 AM [vite] Error when evaluating SSR module /node_modules/__mf__virtual/host_mf_2_page_mf_2_app__loadRemote__remote_mf_1_remote_mf_2_app__loadRemote__.js:
|- ReferenceError: require is not defined
```

see stackblitz for reproduction:
https://stackblitz.com/edit/github-ad1c93-e4uuvj?file=vite.config.ts,src/main.jsx,test.js

**Comment by omridevk** [comment]
run 
`npm run test` 
to see the error

**Comment by gioboa** [comment]
Thanks @omridevk for your feedback. We still need to focus into the SSR story, would you like to tackle the problem?

**Comment by omridevk** [comment]
sure, it was quite easy to resolve this by using 
```
import { createRequire } from 'node:module';
const require = createRequire(import.meta.url);
```
and using
```default export```
instead of the CJS module.exports

**Comment by gioboa** [comment]
@omridevk awesome :clap: I'm looking forward to testing your PR :rocket:

**Comment by RobinBomkamp** [comment]
We are also facing the same problem in SvelteKit. It was working in version 0.2.8. see #163

**Comment by omridevk** [comment]
Yes sorry, had to shift my focus on something else, hopefully will get to this, this week

**Comment by veeramarni** [comment]
I couldn't find a workaround for this issue. I'm testing this with Remix app.

**Comment by omridevk** [comment]
@gioboa  I was able to get the preview working, but I can't seem to be able to run tests, it is running into some memory issues, i'll add the logs later today

**Comment by omridevk** [comment]
`virtualRemotes.ts`

```
import VirtualModule from '../utils/VirtualModule';
import { virtualRuntimeInitStatus } from './virtualRuntimeInitStatus';

const cacheRemoteMap: {
  [remote: string]: VirtualModule;
} = {};
export const LOAD_REMOTE_TAG = '__loadRemote__';
export function getRemoteVirtualModule(remote: string, command: string) {
  if (!cacheRemoteMap[remote]) {
    cacheRemoteMap[remote] = new VirtualModule(remote, LOAD_REMOTE_TAG, '.js');
    cacheRemoteMap[remote].writeSync(generateRemotes(remote, command));
  }
  const virtual = cacheRemoteMap[remote];
  return virtual;
}
const usedRemotesMap: Record<string, Set<string>> = {
  // remote1: {remote1/App, remote1, remote1/Button}
};
export function addUsedRemote(remoteKey: string, remoteModule: string) {
  if (!usedRemotesMap[remoteKey]) usedRemotesMap[remoteKey] = new Set();
  usedRemotesMap[remoteKey].add(remoteModule);
}
export function getUsedRemotesMap() {
  return usedRemotesMap;
}
export function generateRemotes(id: string, command: string) {
  return `
    import { createRequire } from 'node:module';
    const require = createRequire(import.meta.url);
    const {loadRemote} = require("@module-federation/runtime")
    const {initPromise} = require("${virtualRuntimeInitStatus.getImportId()}")
    const res = initPromise.then(_ => loadRemote(${JSON.stringify(id)}))
    const exportModule = ${command !== 'build' ? '/*mf top-level-await placeholder replacement mf*/' : 'await '}initPromise.then(_ => res)
    export default exportModule;
  `;
}

```

**Comment by gioboa** [comment]
@omridevk nice, can you create a PR with your changes please? In this way we can test it in different environments

**Comment by Seifenn** [comment]
Hello guys is there a develompoent on this issue ? because i have tried @omridevk  solution in my app and i dont get the require is not defined message anymore but i'm having problems with the angular dependencies not being loaded. I'm using analog js for SSR wtih angular.

**Comment by gioboa** [comment]
No, there is no progression on this issue. feel free to contribute :wink:

**Comment by Seifenn** [comment]
The package [angular-architects/module-federation](https://github.com/angular-architects/module-federation-plugin) have added ssr support but their package is not compatible with vite.  I'm trying to understand the changes they made and try to reproduce the same approach on this package since i'm having problems with the angular dependecies not exported.

``bash
✘ [ERROR] No matching export in "node_modules/__mf__virtual/host_mf_2_app__loadShare___mf_0_angular_mf_1_core__loadShare__.js" for import "Inject"
    node_modules/@angular/router/fesm2022/router.mjs:8:642:
      8 │ ...ctor, ApplicationRef, InjectFlags, APP_INITIALIZER, SkipSelf, NgModule, Inject, Version } from '@angular/core';
``
      
I'm not sure of the cause of the error but maybe we can follow the changes they have done to make this module support ssr i'm using analog js as my host for testing ssr and the project can be found on github [vite-module-federation-angular-test](https://github.com/Seifenn/vite-module-federation-angular-test) . 

Do you have any suggestions what can be the cause of this error in relation with ssr and module exports ?

**Comment by gioboa** [comment]
> I couldn't find a workaround for this issue. I'm testing this with Remix app.

I did [this example](https://github.com/gioboa/remix-microfrontend-demo)

It's not working yet because this import from the host is never resolved ( no errors ) 😮‍💨
```
const Remote = lazy(
  async () => import('remote/remote-app'),
);
```

Any help is more than welcome.

@RobinBomkamp helped with the SvelteKit example and he added few special edge checks for it here #206 
idk if we need to do the same for Remix

**Comment by gioboa** [comment]
@omridevk I tried the basic reproduction repo and I update the dependencies
```
"@module-federation/vite": "1.2.0",
"vite": "6.0.4"
```
It's working fine

**Comment by GoncaloHit** [comment]
> sure, it was quite easy to resolve this by using
> 
> ```
> import { createRequire } from 'node:module';
> const require = createRequire(import.meta.url);
> ```
> 
> and using `default export` instead of the CJS module.exports

Where should i add this ?
whenever i add the federation plugin to vite config i get the same error as you ?
i am using react SSR, if i remove the federation plugin it works great

**Comment by poenneby** [comment]
I've created a PR based changes proposed by @omridevk. Tests are currently failing, will have a look into why but if you see anything obvious let me know.

**Comment by omridevk** [comment]
> reproduction

I am still able to reproduce it with latest version of vite (6.x) and latest version of module fedration vite plugin

**Comment by omridevk** [comment]
not in the my reproduction link, will try to create a new one with the issue persist.

**Comment by Hunga1** [comment]
We're seeing the same issue using React Router 7 (Framework mode), and have been trying to find a root cause and solution but no luck so far.

Reproduction:
https://github.com/Hunga1/module-federation-react-router-framework-reprex

**Comment by forceuser** [comment]
able to reproduce the issue when using **shared** dependencies in  vite.config
https://stackblitz.com/edit/github-ad1c93-czs3i9p1?file=vite.config.ts,src%2Fmain.jsx,test.js

**Comment by neetigyachahar** [comment]
> We're seeing the same issue using React Router 7 (Framework mode), and have been trying to find a root cause and solution but no luck so far.
> 
> Reproduction: https://github.com/Hunga1/module-federation-react-router-framework-reprex

I have been trying to create a similar app with React Router 7 (framework mode) and Module Federation in Vite but getting similar issues. 
I have created a discussion on React Router repo if someone can help out. https://github.com/remix-run/react-router/discussions/13444

Reproduction repo: https://github.com/neetigyachahar/react-router-vite-module-federation

**Comment by neetigyachahar** [comment]
> > We're seeing the same issue using React Router 7 (Framework mode), and have been trying to find a root cause and solution but no luck so far.
> > Reproduction: https://github.com/Hunga1/module-federation-react-router-framework-reprex
> 
> I have been trying to create a similar app with React Router 7 (framework mode) and Module Federation in Vite but getting similar issues. I have created a discussion on React Router repo if someone can help out. [remix-run/react-router#13444](https://github.com/remix-run/react-router/discussions/13444)
> 
> Reproduction repo: https://github.com/neetigyachahar/react-router-vite-module-federation

```tsx
  resolve: {
    alias: {
      "@module-federation/runtime": path.resolve(
        __dirname,
        "node_modules/@module-federation/runtime/dist/index.esm.mjs"
      ),
    },
  },
```

Adding this manually worked as of now. Not sure if the `@module-federation/vite` plugin needs to be updated for a permanent fix. New to bundlers.

**Comment by ajmeraaxesh** [comment]
@neetigyachahar  did the module federation work for you?
